### PR TITLE
Move BuildRID lists to common props

### PR DESCRIPTION
**Pull request by am11**
Platform matrix testing is in progress, but locally made sure that the `artifacts/package` directory tree on macOS is identical before and after this change.

This PR is simplifying the BuildRID part. It removes `OfficialBuildRID` and `UnofficialBuildRID` (not to be confused with very similar-looking `OfficialBuildID` which is a contract between runtime and arcade for versioning). Also fixes tizen/armel crossgen2 issue at the same time.

cc @ViktorHofer

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@dotnet/runtime-infrastructure and @ericstj as I'm out the next two weeks, can you please help @am11 by reviewing and validating the change? We need official build validation for this change as well. Thanks

**Comment by am11** [comment]
Remaining failures are unrelated to the change (NuGet hangups, EventSource test failures). Marking it as ready for review.

**Review by safern — APPROVED** [review]
Looks reasonable to me, but I would like either @hoyosjs or @jkoritzinsky to take another look as well. 

Also, do we need to test this in an official build?

**Comment by hoyosjs** [comment]
I'll review and kick off an internal build. We should also test this for sourcebuild.

**Comment by am11** [comment]
Thanks. There was a redundant exclusion I have just deleted (it won't change the outcome since RestoreBuildRID is used in a place where ExcludeFromBuildRID is not set).

**Comment by am11** [comment]
@hoyosjs, was the internal build green?

> We should also test this for sourcebuild.

I can see `Build Linux x64 release SourceBuild` is green in `runtime-dev-innerloop`.

**Comment by hoyosjs** [comment]
I did a manual check with non-portable rids for sourcebuild and all looked ok. I also code reviewed and it looks fine. I kicked off one and will report when I see it finish.

**Comment by hoyosjs** [comment]
Looks like there are some repeated assets. 

```
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-arm.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-arm.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-arm64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-arm64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-musl-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-musl-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.linux-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.osx-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'runtime.osx-x64.runtime.native.System.IO.Ports' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-arm.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-arm.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-arm64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-arm64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-musl-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-musl-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.linux-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.osx-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2' 
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21330.2\tools\SdkTasks\PublishBuildAssets.proj(43,5): error : Repeated Asset entry: 'assets/symbols/runtime.osx-x64.runtime.native.System.IO.Ports.6.0.0-preview.7.21360.2.symbols.nupkg' - '6.0.0-preview.7.21360.2'
```

**Comment by am11** [comment]
Can we run that target locally or is it something only run on official build? If there is no secure argument involved please paste the command, I can try to investigate it locally.

**Comment by hoyosjs** [comment]
It means that two different legs produced it. Currently trying to see which two legs produced the same package.

**Comment by hoyosjs** [comment]
Looks like `Libraries Build Linux x64 Release` is producing it and at the same time `Build Linux x64 release AllSubsets_Mono` leg and `Build Linux x64 release AllSubsets_Mono_LLVMJIT` are also producing it and that caused the error - and similarly for all other repeated packages.

The command line for the libraries
- `./build.sh -subset libs  -rc release -configuration Release -ci -arch x64 -framework net6.0  /p:OfficialBuildId=20210710.2 `

And for the allsubsets one:
- `./build.sh -ci -arch x64 -os Linux -s mono+libs+host+packs+mono.mscordbi -c Release /p:OfficialBuildId=20210710.2 /p:CrossBuild=false /p:MonoLLVMUseCxx11Abi=false `

Another difference, we are now suddenly producing these that we weren't: 
    <Package Id="runtime.osx-arm64.Microsoft.NETCore.ILAsm" Version="6.0.0-preview.7.21360.2" />
    <Package Id="runtime.osx-arm64.Microsoft.NETCore.ILDAsm" Version="6.0.0-preview.7.21360.2" />
    <Package Id="runtime.osx-arm64.Microsoft.NETCore.TestHost" Version="6.0.0-preview.7.21360.2" />

File: src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj [hunk 1]
```diff
@@ -4,7 +4,7 @@
     <SkipPackageFileCheck>true</SkipPackageFileCheck>
     <SkipValidatePackage>true</SkipValidatePackage>
   </PropertyGroup>
-  <ItemGroup Condition="'$(PackageTargetRuntime)' == 'linux-arm' or '$(PackageTargetRuntime)' == 'linux-arm64' or '$(PackageTargetRuntime)' == 'linux-x64' or '$(PackageTargetRuntime)' == 'osx-x64' or '$(PackageTargetRuntime)' == 'freebsd-x64'">
+  <ItemGroup Condition="'$(PackageTargetRuntime)' != '' and !$(PackageTargetRuntime.ToLowerInvariant.StartsWith('win'))">
```
Comment (hoyosjs) [inline src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj:7]: This somehow didn't consider ARM64 before.  It also now considers linux-musl variants when it didn't before.

File: src/coreclr/.nuget/Directory.Build.props [hunk 2]
```diff
@@ -66,67 +50,6 @@
...
-    </OfficialBuildRID>
-    <OfficialBuildRID Include="tizen.4.0.0-armel">
-      <Platform>armel</Platform>
-    </OfficialBuildRID>
-    <OfficialBuildRID Include="tizen.5.0.0-armel">
-      <Platform>armel</Platform>
-    </OfficialBuildRID>
-  </ItemGroup>
-  <ItemGroup Condition="$(SupportedPackageOSGroups.Contains(';OSX;'))">
-    <OfficialBuildRID Include="osx-x64" />
```
Comment (hoyosjs) [inline src/coreclr/.nuget/Directory.Build.props:93]: Probably this is why we didn't have arm64 ILAsm

File: src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj (hunk 1 above)
Comment (am11) [reply src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj:7]: ARM64 is present before and after the change. linux-musl is not (that restriction does not seem intentional either). I think this is not the causing the validation error. I think the issue is that we have three ways of including pkgproj, one from src/libraries/libraries-packages, one from .proj file and then again from Directory.Build.props. We should try to simplify it, so transition to SDK/dotnet-pack in future become easier.

In this PR, I have changed complex whitelisting to "require a flat list and exclude" mechanism.

Comment (hoyosjs) [reply src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj:7]: osx-arm64 I mean

Comment (hoyosjs) [reply src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj:7]: I am not saying I am against it. I am just saying we might start seeing assets we didn't see before. I'll all for a single list of support and then projects that need can opt out.

File: src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj [hunk 3]
```diff
@@ -1,4 +1,10 @@
 ﻿<Project DefaultTargets="Build">
+  <PropertyGroup>
+    <BuildingAnOfficialBuildLeg Condition="'$(BuildingAnOfficialBuildLeg)' == '' and '$(OfficialBuildId)' != '' and '$(DotNetBuildFromSource)' != 'true'">true</BuildingAnOfficialBuildLeg>
```
Comment (am11) [inline src/libraries/pkg/runtime.native.System.IO.Ports/runtime.native.System.IO.Ports.pkgproj:3]: @hoyosjs, I had to duplicate this line to use it in the condition in the next line. That appears to have fixed the duplicate System.IO.Ports packages (I was taking local diffs of directory tree on Linux with the command you posted). I have my fingers crossed on the next Official leg run (but first he green CI). :)

**Review by jkoritzinsky — APPROVED** [review]

**Comment by hoyosjs** [comment]
I still see some manifest differences. I'll try later.

**Comment by hoyosjs** [comment]
I checked this again and the following packages are missing now compared to tip of main:

```
<Package Id="runtime.linux-arm.runtime.native.System.IO.Ports" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.linux-arm64.runtime.native.System.IO.Ports" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.linux-x64.runtime.native.System.IO.Ports" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.osx-arm64.Microsoft.NETCore.ILAsm" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.osx-arm64.Microsoft.NETCore.ILDAsm" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.osx-arm64.Microsoft.NETCore.TestHost" Version="6.0.0-preview.7.21362.14" />
<Package Id="runtime.osx-x64.runtime.native.System.IO.Ports" Version="6.0.0-preview.7.21362.14" />
```

but they are referenced by the metapackage runtime.native.System.IO.Ports, and a bunch of unofficial rids that we don't produce packages for:

image

Meanwhile this is the dep list for the one in main:

image

This would probably break things. Also @ericstj not sure who owns ports, but shouldn't the package also exist in main for ARM64 osx? Although to be fair, the musl variants seem not to have one so it might be OK?

**Comment by hoyosjs** [comment]
Yeah, this is pretty sad. But there's some place where things are more coupled than we realize. In this case the official build even was green and so was CI - so we are even missing some deeper validation.

**Comment by ericstj** [comment]
> Also @ericstj Eric St. John FTE not sure who owns ports, but shouldn't the package also exist in main for ARM64 osx? Although to be fair, the musl variants seem not to have one so it might be OK?

@krwq @wfurt and originally worked on it, I think this falls under @dotnet/area-system-io.

**Comment by wfurt** [comment]
as far as I remember we did the extra IO.Ports packages so one can consume them on older versions. While consistency is nice I'm not sure there is real demand since the Serial port is not available for quite a while and it should work when used with current runtime. In some cases like ARM macOS the older versions did not even function on that platform.

**Comment by ericstj** [comment]
> as far as I remember we did the extra IO.Ports packages so one can consume them on older versions

It seems to be used on all frameworks, unix runtimes today, which would include OSX ARM64: https://github.com/dotnet/runtime/blob/b056b7b6a1eca5171e42f697a0b3d3c60d1fc048/src/libraries/System.IO.Ports/src/System.IO.Ports.csproj#L116-L119

It'd be great if that could be removed on latest frameworks.

**Comment by clamp03** [comment]
@am11 Tizen/armel build is still broken. If you don't mind, could I make a new PR to fix Tizen/armel crossgen2 issue based on this PR?

**Comment by am11** [comment]
@clamp03, I am very sorry. I totally forgot about Tizen fix. 🤦
Opened a PR with just those parts: https://github.com/dotnet/runtime/pull/55789.

**Pull request closed without merging**